lib/web_ui/lib/src/engine/html/dom_canvas.dart

```diff
@@ -160,8 +162,75 @@ ui.Color blurColor(ui.Color color, double sigma) {
   return ui.Color((reducedAlpha & 0xff) << 24 | (color.value & 0x00ffffff));
 }
 
+/// When drawing a shape (rect, rrect, circle, etc) in DOM/CSS, the [rect] given
+/// by Flutter needs to be adjusted to what DOM/CSS expect.
+///
+/// This method takes Flutter's [rect] and produces a new rect that can be used
+/// to generate the correct CSS properties to match Flutter's expectations.
+///
+///
+/// Here's what Flutter's given [rect] and [paint.strokeWidth] represent:
+///
+///     top-left    ↓
+///              ┌──↓──────────────────────┐
+///             →→→→x                   x  │←←
+///              │    ┌───────────────┐    │ |
+///              │    │               │    │ |
+///              │    │               │    │ | height
+///              │    │               │    │ |
+///              │    └───────────────┘    │ |
+///              │  x                   x  │←←
+///              └─────────────────────────┘
+/// stroke-width ↑----↑                 ↑
+///                 ↑-------------------↑ width
+///
+///
+///
+/// In the DOM/CSS, here's how the coordinates should look like:
+///
+///   top-left   ↓
+///            →→x─────────────────────────┐
+///              │                         │
+///              │    x───────────────x    │←←
+///              │    │               │    │ |
+///              │    │               │    │ | height
+///              │    │               │    │ |
+///              │    x───────────────x    │←←
+///              │                         │
+///              └─────────────────────────┘
+/// border-width ↑----↑               ↑
+///                   ↑---------------↑ width
+///
+/// As shown in the drawing above, the width/height don't start at the top-left
+/// coordinates. Instead, they start from the inner top-left (inside the border).
+ui.Rect adjustRectForDom(ui.Rect rect, SurfacePaintData paint) {
+  double left = math.min(rect.left, rect.right);
+  double top = math.min(rect.top, rect.bottom);
+  double width = rect.width.abs();
+  double height = rect.height.abs();
+
+  final bool isStroke = paint.style == ui.PaintingStyle.stroke;
+  final double strokeWidth = paint.strokeWidth ?? 0.0;
```

Comment:
Could we remove both of the `if` blocks below, if we did the following ahead of time?

```dart
if (paint.style == ui.PaintingStyle.fill || strokeWidth == 0.0) {
  return rect;
}
```

Reply:
Unfortunately, we can't.

Stroke-related adjustments aren't the only ones needed. We need to also adjust `left` and `top` to avoid flipped rectangles.

E.g. `Rect.fromLTRB(200, 200, 100, 100)` is completely fine in Flutter, but the DOM doesn't like it. If you do `rect.width` on a flipped rectangle, you get a negative value. Setting a negative width/height on a DOM element is equivalent to zero.

`buildDrawRectElement` and various draw operations used to handle it. See this:
https://github.com/flutter/engine/blob/2ced97978624d0ad04508adc8fcb447319270bd1/lib/web_ui/lib/src/engine/html/dom_canvas.dart#L175-L178
and: https://github.com/flutter/engine/blob/2ced97978624d0ad04508adc8fcb447319270bd1/lib/web_ui/lib/src/engine/html/bitmap_canvas.dart#L445-L449

All of that can go away now because rects produced by the new `adjustRectForDom` are never flipped.